File: Frontend/lib/features/notes/domain/models/note_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../utils/note_mime_helper.dart';

enum ResourceType { textNote, externalLink, fileAttachment }

extension ResourceTypeExtension on ResourceType {
  String get value {
    switch (this) {
      case ResourceType.textNote:
        return 'text_note';
      case ResourceType.externalLink:
        return 'external_link';
      case ResourceType.fileAttachment:
        return 'file_attachment';
    }
  }

  String get displayName {
    switch (this) {
      case ResourceType.textNote:
        return 'Text Note';
      case ResourceType.externalLink:
        return 'External Link';
      case ResourceType.fileAttachment:
        return 'File Attachment';
    }
  }

  static ResourceType fromString(String val) {
    switch (val.toLowerCase()) {
      case 'text_note':
      case 'textnote':
        return ResourceType.textNote;
      case 'external_link':
      case 'externallink':
        return ResourceType.externalLink;
      case 'file_attachment':
      case 'fileattachment':
        return ResourceType.fileAttachment;
      default:
        return ResourceType.fileAttachment;
    }
  }
}

enum NoteStatus { draft, published, unpublished, archived }

extension NoteStatusExtension on NoteStatus {
  String get value => name;

  String get displayName {
    switch (this) {
      case NoteStatus.draft:
        return 'Draft';
      case NoteStatus.published:
        return 'Published';
      case NoteStatus.unpublished:
        return 'Unpublished';
      case NoteStatus.archived:
        return 'Archived';
    }
  }

  static NoteStatus fromString(String val) {
    final lower = val.toLowerCase();
    return NoteStatus.values.firstWhere(
      (e) => e.name.toLowerCase() == lower,
      orElse: () => NoteStatus.draft,
    );
  }
}

class NoteModel {
  final String id;
  final String title;
  final String description;
  final String? content;
  final String? externalUrl;
  final ResourceType resourceType;
  
  // File Attachment Metadata
  final String? chapter;
  final String? fileName;
  final String? fileType;
  final int? fileSize;
  final String? fileUrl;
  final String? thumbnailUrl;
  final String? storagePath;
  final String? fileId;
  final String? mimeType;
  final int version;
  
  // Context
  final String subjectId;
  final String sectionId;
  final String courseId;
  final String departmentId;
  final String collegeId;
  final String semesterId;
  final String? academicYearId;
  
  // Ownership
  final String facultyId;
  final String authorUserId;

  // Status
  final NoteStatus status;
  
  // Timestamps
  final DateTime? publishedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  const NoteModel({
    required this.id,
    required this.title,
    required this.description,
    this.content,
    this.externalUrl,
    required this.resourceType,
    this.chapter,
    this.fileName,
    this.fileType,
    this.fileSize,
    this.fileUrl,
    this.thumbnailUrl,
    this.storagePath,
    this.fileId,
    this.mimeType,
    this.version = 1,
    required this.subjectId,
    this.sectionId = '',
    this.courseId = '',
    this.departmentId = '',
    this.collegeId = '',
    required this.semesterId,
    this.academicYearId,
    this.facultyId = '',
    required this.authorUserId,
    this.status = NoteStatus.draft,
    this.publishedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isPreviewable => NoteMimeHelper.isPreviewableFormat(fileType ?? fileName);

  factory NoteModel.fromJson(Map<String, dynamic> json) {
    return NoteModel(
      id: (json['id'] ?? json['_id'] ?? '').toString(),
      title: (json['title'] ?? '').toString(),
      description: (json['description'] ?? '').toString(),
      content: json['content'] as String?,
      externalUrl: json['externalUrl'] as String?,
      resourceType: json['resourceType'] != null
          ? ResourceTypeExtension.fromString(json['resourceType'].toString())
          : ResourceType.fileAttachment,
      chapter: json['chapter'] as String?,
      fileName: (json['fileName'] ?? json['originalFileName']) as String?,
      fileType: (json['fileType'] ?? json['fileExtension'] ?? json['mimeType']) as String?,
      fileSize: json['fileSize'] is int
          ? json['fileSize'] as int
          : (json['fileSize'] != null ? int.tryParse(json['fileSize'].toString()) : null),
      fileUrl: json['fileUrl'] as String?,
      thumbnailUrl: json['thumbnailUrl'] as String?,
      storagePath: (json['storagePath'] ?? json['storageKey']) as String?,
      fileId: json['fileId'] as String?,
      mimeType: json['mimeType'] as String?,
      version: json['version'] is int
          ? json['version'] as int
          : (json['version'] != null ? int.tryParse(json['version'].toString()) ?? 1 : 1),
      subjectId: (json['subjectId'] ?? '').toString(),
      sectionId: (json['sectionId'] ?? '').toString(),
      courseId: (json['courseId'] ?? '').toString(),
      departmentId: (json['departmentId'] ?? '').toString(),
      collegeId: (json['collegeId'] ?? '').toString(),
      semesterId: (json['semesterId'] ?? '').toString(),
      academicYearId: json['academicYearId']?.toString(),
      facultyId: (json['facultyId'] ?? '').toString(),
      authorUserId: (json['authorUserId'] ?? '').toString(),
      status: json['status'] != null
          ? NoteStatusExtension.fromString(json['status'].toString())
          : NoteStatus.draft,
      publishedAt: json['publishedAt'] != null ? _parseDate(json['publishedAt']) : null,
      createdAt: json['createdAt'] != null ? _parseDate(json['createdAt'])! : DateTime.now(),
      updatedAt: json['updatedAt'] != null ? _parseDate(json['updatedAt'])! : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'content': content,
      'externalUrl': externalUrl,
      'resourceType': resourceType.value,
      'chapter': chapter,
      'fileName': fileName,
      'fileType': fileType,
      'fileSize': fileSize,
      'fileUrl': fileUrl,
      'thumbnailUrl': thumbnailUrl,
      'storagePath': storagePath,
      'fileId': fileId,
      'mimeType': mimeType,
      'version': version,
      'subjectId': subjectId,
      'sectionId': sectionId,
      'courseId': courseId,
      'departmentId': departmentId,
      'collegeId': collegeId,
      'semesterId': semesterId,
      if (academicYearId != null) 'academicYearId': academicYearId,
      'facultyId': facultyId,
      'authorUserId': authorUserId,
      'status': status.value,
      'publishedAt': publishedAt?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  static DateTime? _parseDate(dynamic date) {
    if (date == null) return null;
    if (date is Timestamp) return date.toDate();
    if (date is String) return DateTime.tryParse(date);
    return null;
  }

  NoteModel copyWith({
    String? id,
    String? title,
    String? description,
    String? content,
    String? externalUrl,
    ResourceType? resourceType,
    String? chapter,
    String? fileName,
    String? fileType,
    int? fileSize,
    String? fileUrl,
    String? thumbnailUrl,
    String? storagePath,
    String? fileId,
    String? mimeType,
    int? version,
    String? subjectId,
    String? sectionId,
    String? courseId,
    String? departmentId,
    String? collegeId,
    String? semesterId,
    String? academicYearId,
    String? facultyId,
    String? authorUserId,
    NoteStatus? status,
    DateTime? publishedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return NoteModel(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      content: content ?? this.content,
      externalUrl: externalUrl ?? this.externalUrl,
      resourceType: resourceType ?? this.resourceType,
      chapter: chapter ?? this.chapter,
      fileName: fileName ?? this.fileName,
      fileType: fileType ?? this.fileType,
      fileSize: fileSize ?? this.fileSize,
      fileUrl: fileUrl ?? this.fileUrl,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      storagePath: storagePath ?? this.storagePath,
      fileId: fileId ?? this.fileId,
      mimeType: mimeType ?? this.mimeType,
      version: version ?? this.version,
      subjectId: subjectId ?? this.subjectId,
      sectionId: sectionId ?? this.sectionId,
      courseId: courseId ?? this.courseId,
      departmentId: departmentId ?? this.departmentId,
      collegeId: collegeId ?? this.collegeId,
      semesterId: semesterId ?? this.semesterId,
      academicYearId: academicYearId ?? this.academicYearId,
      facultyId: facultyId ?? this.facultyId,
      authorUserId: authorUserId ?? this.authorUserId,
      status: status ?? this.status,
      publishedAt: publishedAt ?? this.publishedAt,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

File: Backend/src/models/note.model.ts
import mongoose, { Document, Schema } from 'mongoose';
import {
  NoteStatus,
  NoteType,
  NoteLifecycleState,
  NoteVisibility,
  ResourceType,
} from '../constants/status';

export interface INoteVersion {
  version: number;
  storageKey: string;
  fileId: string;
  fileName: string;
  fileSize: number;
  mimeType: string;
  fileUrl?: string;
  uploadedAt: Date;
}

export interface INote extends Document {
  collegeId: mongoose.Types.ObjectId;
  departmentId: mongoose.Types.ObjectId;
  courseId: mongoose.Types.ObjectId;
  academicYearId?: mongoose.Types.ObjectId;
  semesterId: mongoose.Types.ObjectId;
  sectionId?: mongoose.Types.ObjectId;
  subjectId: mongoose.Types.ObjectId;
  facultyId?: mongoose.Types.ObjectId;
  authorUserId: mongoose.Types.ObjectId;
  title: string;
  description: string;
  noteType: NoteType;
  visibility: NoteVisibility;
  lifecycleState: NoteLifecycleState;
  status: NoteStatus;
  resourceType: ResourceType;
  storageProvider: string;
  originalFileName: string;
  storageKey: string;
  fileId: string;
  fileUrl?: string;
  thumbnailUrl?: string;
  mimeType: string;
  fileExtension: string;
  fileSize: number;
  version: number;
  previousVersions: INoteVersion[];
  chapter?: string;
  publishedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const NoteVersionSchema = new Schema<INoteVersion>(
  {
    version: { type: Number, required: true },
    storageKey: { type: String, required: true },
    fileId: { type: String, required: true },
    fileName: { type: String, required: true },
    fileSize: { type: Number, required: true },
    mimeType: { type: String, required: true },
    fileUrl: { type: String, default: null },
    uploadedAt: { type: Date, default: Date.now },
  },
  { _id: false }
);

const NoteSchema = new Schema<INote>(
  {
    collegeId: { type: Schema.Types.ObjectId, ref: 'College', required: true, index: true },
    departmentId: { type: Schema.Types.ObjectId, ref: 'Department', required: true, index: true },
    courseId: { type: Schema.Types.ObjectId, ref: 'Course', required: true, index: true },
    academicYearId: { type: Schema.Types.ObjectId, ref: 'AcademicYear', default: null },
    semesterId: { type: Schema.Types.ObjectId, ref: 'Semester', required: true, index: true },
    sectionId: { type: Schema.Types.ObjectId, ref: 'Section', default: null, index: true },
    subjectId: { type: Schema.Types.ObjectId, ref: 'Subject', required: true, index: true },
    facultyId: { type: Schema.Types.ObjectId, ref: 'Faculty', default: null, index: true },
    authorUserId: { type: Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    title: { type: String, required: true, trim: true },
    description: { type: String, default: '', trim: true },
    noteType: {
      type: String,
      enum: Object.values(NoteType),
      default: NoteType.STUDY_MATERIAL,
    },
    visibility: {
      type: String,
      enum: Object.values(NoteVisibility),
      default: NoteVisibility.PUBLIC,
    },
    lifecycleState: {
      type: String,
      enum: Object.values(NoteLifecycleState),
      default: NoteLifecycleState.PENDING_UPLOAD,
      index: true,
    },
    status: {
      type: String,
      enum: Object.values(NoteStatus),
      default: NoteStatus.DRAFT,
      index: true,
    },
    resourceType: {
      type: String,
      enum: Object.values(ResourceType),
      default: ResourceType.FILE_ATTACHMENT,
    },
    storageProvider: {
      type: String,
      default: 'imagekit',
      required: true,
    },
    originalFileName: { type: String, required: true, trim: true },
    storageKey: { type: String, required: true, unique: true },
    fileId: { type: String, required: true },
    fileUrl: { type: String, default: null },
    thumbnailUrl: { type: String, default: null },
    mimeType: { type: String, required: true },
    fileExtension: { type: String, required: true },
    fileSize: { type: Number, required: true, min: 1 },
    version: { type: Number, default: 1 },
    previousVersions: [NoteVersionSchema],
    chapter: { type: String, default: null },
    publishedAt: { type: Date, default: null },
  },
  {
    timestamps: true,
    toJSON: {
      virtuals: true,
      transform: (_: unknown, ret: Record<string, any>) => {
        ret.id = ret._id?.toString();
        ret.collegeId = ret.collegeId?.toString();
        ret.departmentId = ret.departmentId?.toString();
        ret.courseId = ret.courseId?.toString();
        if (ret.academicYearId) ret.academicYearId = ret.academicYearId.toString();
        ret.semesterId = ret.semesterId?.toString();
        if (ret.sectionId) ret.sectionId = ret.sectionId.toString();
        ret.subjectId = ret.subjectId?.toString();
        if (ret.facultyId) ret.facultyId = ret.facultyId.toString();
        ret.authorUserId = ret.authorUserId?.toString();
        delete ret.__v;
        return ret;
      },
    },
  }
);

NoteSchema.index({ collegeId: 1, subjectId: 1, lifecycleState: 1 });
NoteSchema.index({ collegeId: 1, semesterId: 1, subjectId: 1 });
NoteSchema.index({ collegeId: 1, authorUserId: 1 });
NoteSchema.index({ collegeId: 1, status: 1 });

export const Note = mongoose.model<INote>('Note', NoteSchema);
